streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java

```diff
@@ -0,0 +1,341 @@
+        this.produced = produced;
+        this.consumed = consumed;
+        this.materialized = materialized;
+        this.processorSupplier = processorSupplier;
+        this.storeSupplier = storeSupplier;
+        this.storeBuilder = storeBuilder;
+        this.kTableSource = kTableSource;
+        this.storeNames = storeNames;
+        this.connectProcessorName = connectProcessorName;
+        this.parentNames = parentNames;
+        this.sourceName = sourceName;
+        if (this.materialized != null) {
+            this.materializedInternal = new MaterializedInternal(materialized);
+        }
+
+        if (produced != null) {
+            producedInternal = new ProducedInternal<>(produced);
+        }
+
+        if (consumed != null) {
+            consumedInternal = new ConsumedInternal<>(consumed);
+        }
+    }
+
+    public Collection<String> getSourceTopics() {
+        final Set<String> copy = new HashSet<>(sourceTopics);
+        return Collections.unmodifiableSet(copy);
+    }
+
+    public String[] getSourceTopicArray() {
+        if (sourceTopics == null) {
+            return new String[]{};
+        }
+        final String[] topics = new String[sourceTopics.size()];
+        sourceTopics.toArray(topics);
+        return topics;
+    }
+
+    public String getSinkTopic() {
+        return sinkTopic;
+    }
+
+    public Pattern getSourcePattern() {
+        return sourcePattern;
+    }
+
+    public Produced<K, V> getProduced() {
+        return produced;
+    }
+
+    public Consumed<K, V> getConsumed() {
+        return consumed;
+    }
+
+    public Serde<K> producedKeySerde() {
+        return producedInternal.keySerde();
+    }
+
+    public Serde<V> producedValueSerde() {
+        return producedInternal.valueSerde();
+    }
+
+    public StreamPartitioner<? super K, ? super   V> streamPartitioner() {
+
+        final Serializer<K> keySerializer = producedInternal.keySerde() == null ? null : producedInternal.keySerde().serializer();
+        StreamPartitioner<? super K, ? super V> partitioner = producedInternal.streamPartitioner();
+        if (partitioner == null && keySerializer != null && keySerializer instanceof WindowedSerializer) {
+            final WindowedSerializer<Object> windowedSerializer = (WindowedSerializer<Object>) keySerializer;
+            partitioner = (StreamPartitioner<K, V>) new WindowedStreamPartitioner<Object, V>(sinkTopic, windowedSerializer);
+
+        }
+        return partitioner;
+    }
+
+    public Serde<K> consumedKeySerde() {
+        return consumedInternal.keySerde();
+    }
+
+    public Topology.AutoOffsetReset getConsumedResetPolicy() {
+        return consumedInternal.offsetResetPolicy();
+    }
+
+    public TimestampExtractor getConsumedTimestampExtractor() {
+        return consumedInternal.timestampExtractor();
+    }
+
+    public Serde<V> consumedValueSerde() {
+        return consumedInternal.valueSerde();
+    }
+
+    public Materialized<K, V, S> getMaterialized() {
```

Comment:
This function and its corresponding field looks redundant to me: the only callers of this function is checking for null, and when requesting the reference the other `getMaterializedInternal` is used. We can instead only keep the `getMaterializedInternal` and its corresponding fields.

Ditto for produced and consumed.

Reply:
Ack will include in refactoring for comments.